### PR TITLE
Add new Encodable render method 

### PR DESCRIPTION
**Pull request by DunnCoding**
## Description
This PR: 
* Adds a new Codable render method to the Kitura-TemplateEngine protocol.
* Moves to Swift 4 support only

## Motivation and Context
See this [proposal](https://github.com/IBM-Swift/evolution/pull/14)

### CODE REVIEW

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 1]
```diff
@@ -87,6 +118,15 @@ extension TemplateEngine {
                        options: RenderingOptions, templateName: String) throws -> String {
         return try render(filePath: filePath, context: context, options: options)
     }
+    
+    //Implementation of Codable render for TemplateEngines that do not implement it
+    public func render<T: Encodable>(filePath: String, with value: T, forKey key: String?,
+                                   options: RenderingOptions, templateName: String) throws -> String {
+        if let contextKey = key {
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]:
```swift
guard let key = key else {
    throw TemplateEngineError.needToImplementForUseWithoutKey
}
return try render(filePath: filePath, context: [key: value], options: options)
```

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 2]
```diff
@@ -21,6 +21,22 @@ public struct NullRenderingOptions: RenderingOptions {
     public init() {}
 }
 
+public enum TemplateEngineError: Swift.Error {
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Doc comments needed.

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 3]
```diff
@@ -21,6 +21,22 @@ public struct NullRenderingOptions: RenderingOptions {
...
+    case needToImplementForUseWithoutKey
+    case unableToCastJSONToDict(Any)
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: This seems to be unused?

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 4]
```diff
@@ -21,6 +21,22 @@ public struct NullRenderingOptions: RenderingOptions {
...
+}
+
+extension TemplateEngineError: CustomStringConvertible {
+    public var description: String {
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Doc comments needed.

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 5]
```diff
@@ -21,6 +21,22 @@ public struct NullRenderingOptions: RenderingOptions {
...
+        switch self {
+        case .needToImplementForUseWithoutKey:
+            return "This method requires implemetation for use when providing a Codable value without a `forKey` value."
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: implemetation/implementation

File: Sources/KituraTemplateEngine/TemplateEngine.swift (hunk 3 above)
Comment (DunnCoding) [reply Sources/KituraTemplateEngine/TemplateEngine.swift]: `unableToCastJSONToDict ` is used by KituraStencil and KituraMustache. 
I felt as it's used by both I should just add it here where it's available to both.

File: Package.swift [hunk 6]
```diff
@@ -17,5 +20,19 @@
 import PackageDescription
 
 let package = Package(
-    name: "Kitura-TemplateEngine"
+    name: "Kitura-TemplateEngine",
+    products: [
+        // Products define the executables and libraries produced by a package, and make them visible to other packages.
```
Comment (ianpartridge) [inline Package.swift]: I think we can delete this line - it's just boilerplate generated by SPM when you do `swift package init`.

File: Package.swift [hunk 7]
```diff
@@ -17,5 +20,19 @@
...
+        .library(
+            name: "KituraTemplateEngine",
+            targets: ["KituraTemplateEngine"]
+        )
+    ],
+    targets: [
+        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
```
Comment (ianpartridge) [inline Package.swift]: Same for these two.

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 8]
```diff
@@ -20,6 +20,26 @@ public protocol RenderingOptions {}
 public struct NullRenderingOptions: RenderingOptions {
     public init() {}
 }
+/// Template Engine Error
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Blank line above this one, please.

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 9]
```diff
@@ -20,6 +20,26 @@ public protocol RenderingOptions {}
...
+/// case: needToImplementForUseWithoutKey - For when users attempt to use default Codable API without providing a `forKey`.
+/// case: unableToCastJSONToDict(Any) - Used by template engines that conform to Kitura-TemplateEngine when failing to cast
+///                                     the JSONSerialisation result to a [String: Any]
+public enum TemplateEngineError: Swift.Error {
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Did you try using the `struct` approach like in `RequestError`?

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 10]
```diff
@@ -20,6 +20,26 @@ public protocol RenderingOptions {}
...
+    case needToImplementForUseWithoutKey
+    case unableToCastJSONToDict(Any)
+}
+
+/// Extends Template Engine Error to provide user readable descriptions strings when errors are thrown.
+extension TemplateEngineError: CustomStringConvertible {
+    public var description: String {
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Doc comment needed.

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 11]
```diff
@@ -20,6 +20,26 @@ public protocol RenderingOptions {}
...
+        switch self {
+        case .needToImplementForUseWithoutKey: 
+            return "This method requires implementation for use when providing a Codable value without a `forKey` value."
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Sorry, I don't understand this message - can we reword/clarify?

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 12]
```diff
@@ -87,6 +122,16 @@ extension TemplateEngine {
                        options: RenderingOptions, templateName: String) throws -> String {
         return try render(filePath: filePath, context: context, options: options)
     }
+    
+    //Implementation of Codable render for TemplateEngines that do not implement it
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Space after `//`. Also the other comments in this file say `implementation` so please update those too.

**Review by ianpartridge — CHANGES_REQUESTED** [review]

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 13]
```diff
@@ -66,6 +66,21 @@ public protocol TemplateEngine {
     func render(filePath: String, context: [String: Any],
                 options: RenderingOptions, templateName: String) throws -> String
 
+    /// Take a template file and a Codable type and generate the content to be sent back to the client.
+    ///
+    /// - Parameter filePath: The path of the template file to use when generating
+    ///                      the content.
+    /// - Parameter with: A value that conforms to Codable which is used to generate the content.
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Codable -> Encodable

File: Sources/KituraTemplateEngine/TemplateEngine.swift [hunk 14]
```diff
@@ -66,6 +66,21 @@ public protocol TemplateEngine {
...
+    ///
+    /// - Parameter forKey: A value used to match the Codable values to the correct variable in a template file.
```
Comment (ianpartridge) [inline Sources/KituraTemplateEngine/TemplateEngine.swift]: Codable -> Encodable

**Review by ianpartridge — CHANGES_REQUESTED** [review]

**Review by ianpartridge — APPROVED** [review]

**Review by Andrew-Lees11 — APPROVED** [review]
LGTM

**Pull request merged**